packages/yarnpkg-pnp/sources/esm-loader/loader.ts

```diff
@@ -0,0 +1,192 @@
+import {NativePath, PortablePath}          from '@yarnpkg/fslib';
+import {parse, init}                       from 'cjs-module-lexer';
+import fs                                  from 'fs';
+import moduleExports                       from 'module';
+import path                                from 'path';
+import {fileURLToPath, pathToFileURL, URL} from 'url';
+
+import * as nodeUtils                      from '../loader/nodeUtils';
+import {PnpApi}                            from '../types';
+
+function isValidURL(str: string) {
+  try {
+    new URL(str);
+    return true;
+  } catch {
+    return false;
+  }
+}
+
+const builtins = new Set([...moduleExports.builtinModules]);
+
+const pathRegExp = /^(?![a-zA-Z]:[\\/]|\\\\|\.{0,2}(?:\/|$))((?:node:)?(?:@[^/]+\/)?[^/]+)\/*(.*|)$/;
+
+async function exists(path: string) {
+  try {
+    await fs.promises.access(path, fs.constants.R_OK);
+    return true;
+  } catch { }
+  return false;
+}
+
+export async function resolve(
+  originalSpecifier: string,
+  context: any,
+  defaultResolver: any
+) {
+  const {findPnpApi} = (moduleExports as unknown as { findPnpApi?: (path: NativePath) => null | PnpApi });
+  if (!findPnpApi)
+    return defaultResolver(originalSpecifier, context, defaultResolver);
+
+  let specifier = originalSpecifier;
+  let validURL: boolean | undefined;
+  if (builtins.has(specifier) || (validURL = isValidURL(specifier))) {
+    if (!validURL || pathToFileURL(specifier).protocol !== `file:`) {
+      return defaultResolver(originalSpecifier, context, defaultResolver);
+    } else {
+      specifier = fileURLToPath(specifier);
+    }
+  }
+
+  const {parentURL, conditions = []} = context;
+
+  const issuer = parentURL ? fileURLToPath(parentURL) : process.cwd();
+
+  const pnpapi = findPnpApi(issuer) ?? (validURL ? findPnpApi(specifier) : null);
```

Comment:
I don't understand the reason for the fallback, even though I'm sure there's one. A comment would be beneficial.

Reply:
It's for when the `specifier` (a file path) is outside of the current project, for example when running `yarn dlx`